### PR TITLE
Allow csv_collection to pass options to CSV parser

### PR DESCRIPTION
**Pull request by bdewater**
This allows tasks to pass [options for Ruby's CSV parser](https://ruby-doc.org/3.3.0/stdlibs/csv/CSV.html#class-CSV-label-Options+for+Parsing) which can do useful things like automatically coercing numbers and dates, stripping whitespace, etc. While those things could also be implemented in a task itself, other things like skipping over certain non-CSV (comment) lines or defining the separators prevent the parser from raising errors.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -201,6 +201,37 @@ The files uploaded to your Active Storage service provider will be renamed to
 include an ISO 8601 timestamp and the Task name in snake case format. The CSV is
 expected to have a trailing newline at the end of the file.
 
+#### CSV options
+
+Tasks can pass [options for Ruby's CSV parser][csv-parse-options] by adding a
+`csv_options` hash to the `csv_collection` macro:
```
Comment (etiennebarrie) [inline README.md:207]: I think this should be "... by adding keyword arguments to `csv_collection`."

**Review by etiennebarrie — COMMENTED** [review]
I approve of the idea but I don't like having to pass a keyword arg with a hash for the options. I think we could just grab all remaining keyword args and send them to CSV instead.

**Comment by bdewater** [comment]
Thanks for the feedback @etiennebarrie, PTAL!

File: app/models/maintenance_tasks/batch_csv_collection_builder.rb [hunk 2]
```diff
@@ -12,16 +12,17 @@ class BatchCsvCollectionBuilder < CsvCollectionBuilder
...
-      super()
+      super(csv_options)
     end
 
     # Defines the collection to be iterated over, based on the provided CSV.
     # Includes the CSV and the batch size.
     def collection(task)
       BatchCsv.new(
-        csv: CSV.new(task.csv_content, headers: true),
+        csv: CSV.new(task.csv_content, **@csv_options.merge(headers: true)),
```
[the remark this answers is not in the document]
Comment (bdewater) [reply app/models/maintenance_tasks/batch_csv_collection_builder.rb]: Yes much better, should've thought of this!

**Review by etiennebarrie — APPROVED** [review]
Looks good to me. @nvasilevski @adrianna-chang-shopify you're OK with the API? (folding CSV kwargs into `csv_collection` directly)

**Review by nvasilevski — APPROVED** [review]
Looks clean to me 👍

**Comment by bdewater** [comment]
I noticed I forgot to update the YARD docs after incorporating the PR feedback. Give me a few moments to fix :)

**Review by adrianna-chang-shopify — COMMENTED** [review]
Looks like the CSV count test is failing, but this looks good to me otherwise 👍

**Comment by D-system** [comment]
My current workaround is this patch:
```ruby
Rails.application.reloader.to_prepare do
  MaintenanceTasks::CsvCollectionBuilder.class_eval do
    def collection(task)
      CSV.new(task.csv_content, headers: true, encoding: Encoding::UTF_8)
    end
  end
end
```

For the ASCII only CSV, even with the `Encoding::BINARY` is working puzzled me. And binary is an alias for ASCII-8BIT (confirmed with `Encoding.aliases`)
https://github.com/ruby/ruby/blob/b3c59370ca20fcde2fde5a63361550c04ca44be3/encoding.c#L1848

At the beginning, I was thinking to do a encoding method on the `CsvCollectionBuilder` to return UTF_8 that could be overwritten per task.
With the `csv_collection` from this PR, I wonder if adding an default option wouldn't be preferable (same as `headers` https://github.com/Shopify/maintenance_tasks/pull/1001/files#diff-b8b2c5f21a247ff7e047ca62a76f3879a72e963e7bce8a4a4c538cddd465ce2aR81)

```ruby
csv_options[:encoding] = Encoding::UTF_8 unless csv_options.key?(:encoding)
```

As UTF_8 is 100% compatible with ASCII it should not be a breaking change. I'm not aware of the specification of the 8bit version of ASCII (the `ASCII-8BIT`).


Should I go ahead with the new option for `csv_options`?

**Comment by bdewater** [comment]
> For the ASCII only CSV, even with the `Encoding::BINARY` is working puzzled me.

It's [in the docs](https://ruby-doc.org/3.3.0/Encoding.html) 🙂 "Encoding::ASCII_8BIT is a special-purpose encoding that is usually used for a string of bytes, not a string of characters. But as the name indicates, its characters in the ASCII range are considered as ASCII characters."

> Should I go ahead with the new option for `csv_options`?

FWIW I was also thinking of a default option in that way, but @etiennebarrie has a good point about `Encoding.default_external` instead of hard coding UTF-8, so it's the same encoding the Ruby uses by default for CSV, File, etc (see above link).

**Comment by D-system** [comment]
Good point, I missed that part, let's use the defaults `Encoding.default_external`

**Pull request merged**